### ISSUE TITLE
Doesn't work on windows + Julia 0.7

### ISSUE THREAD

**Issue by Datseris** [JuliaLang/IJulia.jl#693]
Kernel opens but operations hang indefinitely (e.g. doing `2+2` hangs, never returns).

**Comment by stevengj** [comment]
See https://github.com/JuliaLang/IJulia.jl#debugging-ijulia-problems to get more output.

**Comment by Datseris** [comment]
Hi Steven, I did this but it didn't help me...

You see, what happens is that the computation freezes indefinitely: it *does not* return an error. (doesn't return anything)

![image](https://user-images.githubusercontent.com/19669089/43370094-1e00a96e-9379-11e8-9164-dd3ff9397332.png)

This is the picture I see (forever) when I try to run a computation. User `Marco Concetto Rudilosso` reported same behavior.

**Comment by Datseris** [comment]
Pressing the button `interrupt the kernel` also does nothing.

**Comment by stevengj** [comment]
You have to look in the terminal where you launched `jupyter` to see the [debugging info](https://github.com/JuliaLang/IJulia.jl#debugging-ijulia-problems)

**Comment by Datseris** [comment]
I don't understand.

1. I have done what is said in the documentation paragraph you point (changing the source code).
2. I open a new `notebook()`. I try to run `1+1` and it never returns.
3. I press `interrupt the kernel`, in hope of seeing the "errors" mentioned in the documentation paragraph, but I see nothing because the Julia kernel never dies.
4. I go into the REPL that I typed `notebook()` and I press Ctrl+C to terminate process. All I see is
```
julia> using IJulia

julia> notebook()
WARNING: Force throwing a SIGINT
```
The kernel doesn't actually terminate. I go back to the Jupyter page and it says "The kernel appears to have died. It will restart automatically." or whatever.

And that is all. After that, everything freezes forever **including the REPL**. I can't do anything, no matter how many times I press Cntrl+C I still simply get the message `The kernel appears to have died. It will restart automatically.`

**Comment by stevengj** [comment]
I just tried 0.7 on a Windows machine, and I can reproduce your hang.   However, if I turn on debugging output, then it works!  Arghh.

Turning on debugging output gives me a lot of output (as expeced) in Windows if I run `jupyter notebook` directly in the Windows command shell, but gives me no output with `notebook()`, which is also annoying.

Perhaps you are editing the wrong files?  On `master` (type `dev IJulia` at the package prompt), I made it a bit easier to turn on debugging output: just set `ENV["IJULIA_DEBUG"]=true` and then re-build IJulia (type `build IJulia` at the package prompt), then re-launch the notebook.

**Comment by Datseris** [comment]
Thansk a lot for the effort! I'll try it tomorrow first thing in the morning!

**Comment by cherichy** [comment]
Same issue. 
I have tried to use IJulia in Debug mode and rebuild the package, but it still won't work.
I even try to use the Jupyter qtconsole, for a simple expression like “1+1”, qtconsole can give the right result. But if I using some pkg such as "using PyPlot", the expression won't return and get stuck like what it shows in the notebook. However, in the notebook, the simplest "1+1" will not return.

**Comment by Datseris** [comment]
@stevengj :( It did not work... I followed exactly the steps of your last comment.
Then I opened up PowerShell, and did:
```

PS C:\Users\datseris> jupyter notebook
```
and then opened a new Julia0.7-beta2 notebook. Pressed 1+1, and still it hang forever. 
here is what is shown in the powershell:
```
[I 14:39:50.947 NotebookApp] Serving notebooks from local directory: C:\Users\datseris
[I 14:39:50.947 NotebookApp] 0 active kernels
[I 14:39:50.947 NotebookApp] The Jupyter Notebook is running at:
[I 14:39:50.947 NotebookApp] http://localhost:8888/?token=32772aedea2e8aac474b1da19f1907bffa91f1f1cef4f66b
[I 14:39:50.947 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 14:39:50.957 NotebookApp]

    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=32772aedea2e8aac474b1da19f1907bffa91f1f1cef4f66b
[I 14:39:51.582 NotebookApp] Accepting one-time-token-authenticated connection from ::1
[I 14:40:19.612 NotebookApp] Creating new notebook in
[I 14:40:23.494 NotebookApp] Kernel started: 6c484567-c4db-409e-a2dc-edc95cff2588
┌ Warning: `srand` is deprecated, use `Random.seed!` instead.
│   caller = __init__() at init.jl:22
└ @ IJulia C:\Users\datseris\.julia\dev\IJulia\src\init.jl:22
PROFILE = Dict{String,Any}("key"=>"0cb586b5-930376e1e5de85d80147afc5","transport"=>"tcp","signature_scheme"=>"hmac-sha25
6","ip"=>"127.0.0.1","hb_port"=>51863,"control_port"=>51862,"shell_port"=>51859,"stdin_port"=>51861,"kernel_name"=>"juli
a-0.7","iopub_port"=>51860)
14:40:29(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict("username"=>"jlkernel","session"=>"UUID(\"c9634273-e7e6-4710-a2b1-2fdead814e0b\")"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"68ec7878-3e0a-4453-b184-d21358557c86\")","msg_type"=>"status","username"=>
"jlkernel","date"=>2018-07-30T14:40:29.339,"version"=>"5.0","session"=>"UUID(\"c9634273-e7e6-4710-a2b1-2fdead814e0b\")")
,
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"starting")
}
Starting kernel event loops.
14:40:31(write task): got msg part ���{&
14:40:31(write task): got msg part <IDS|MSG>
14:40:31(write task): RECEIVED IPython Msg [ idents ���{& ] {
  parent_header = Dict{String,Any}(),
  header = Dict{String,Any}("msg_id"=>"c6ecf981-f30a9fcf1146cdb1abe65b9d","msg_type"=>"kernel_info_request","username"=>
"username","date"=>"2018-07-30T12:40:23.524000Z","version"=>"5.3","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}()
}
14:40:31(write task): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"c6ecf981-f30a9fcf1146cdb1abe65b9d","msg_type"=>"kernel_info_request","user
name"=>"username","date"=>"2018-07-30T12:40:23.524000Z","version"=>"5.3","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"74460e21-4931-48a7-94c2-d553a7db7ffb\")","msg_type"=>"status","username"=>
"username","date"=>2018-07-30T14:40:31.855,"version"=>"5.0","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"busy")
}
14:40:32(write task): SENDING IPython Msg [ idents ���{& ] {
  parent_header = Dict{String,Any}("msg_id"=>"c6ecf981-f30a9fcf1146cdb1abe65b9d","msg_type"=>"kernel_info_request","user
name"=>"username","date"=>"2018-07-30T12:40:23.524000Z","version"=>"5.3","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"def4ad43-837a-4ec8-a168-60ce46a7f3de\")","msg_type"=>"kernel_info_reply","
username"=>"username","date"=>2018-07-30T14:40:32.138,"version"=>"5.0","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("help_links"=>Dict{String,String}[Dict("text"=>"Julia Home Page","url"=>"http://julialang.o
rg/"), Dict("text"=>"Julia Documentation","url"=>"http://docs.julialang.org/"), Dict("text"=>"Julia Packages","url"=>"ht
tp://pkg.julialang.org/")],"language_info"=>Dict("file_extension"=>".jl","name"=>"julia","mimetype"=>"application/julia"
,"version"=>"0.7.0"),"implementation"=>"ijulia","banner"=>"Julia: A fresh approach to technical computing.","protocol_ve
rsion"=>"5.0")
}
[I 14:40:32.434 NotebookApp] Adapting to protocol v5.0 for kernel 6c484567-c4db-409e-a2dc-edc95cff2588
14:40:32(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"c6ecf981-f30a9fcf1146cdb1abe65b9d","msg_type"=>"kernel_info_request","user
name"=>"username","date"=>"2018-07-30T12:40:23.524000Z","version"=>"5.3","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"b6121427-057e-4411-9c0c-b044dbdee611\")","msg_type"=>"status","username"=>
"username","date"=>2018-07-30T14:40:32.516,"version"=>"5.0","session"=>"6157ca63f29a45339b604f2cf772d2b8"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"idle")
}
```
What did I do wrong? I `dev`ed IJulia, and then set the ENV variable and then `build IJulia` and then powershell -> notebook.

**Comment by stevengj** [comment]
I don't know why turning on debugging output made the problem go away for me — it makes debugging a lot harder, that's for sure.

**Comment by Datseris** [comment]
@stevengj could it be that we need masters on some dependency of IJulia? In general, if possible, I would really like to be able to use Jupyter before JuliaCon because I wanted to publish a video and talk about it with people there.

If you have any suggestions whatsoever please let me know and I will try them! Even if they are hard!

**Comment by stevengj** [comment]
I don't know, according to the log it is not even receiving the `execute_request` message from executing the cell.

A random thought — you could try changing [this line](https://github.com/JuliaLang/IJulia.jl/blob/7fcd0ef62f8a0ab06cd28b21dcd50aa5ef54aa18/src/handlers.jl#L178) to `"5.3"` in case there is something wrong with Jupyter's protocol adapter for 5.0.

**Comment by stevengj** [comment]
The `got msg part ���{&` looks weird...

**Comment by stevengj** [comment]
Hmm, I wonder if it due somehow to the recent switch in Conda.jl to default to Python 3?   If `Conda.MINICONDA_VERSION` is `"3"`, you could try [switching it to Python 2](https://github.com/JuliaPy/Conda.jl#using-python-2).

**Comment by stevengj** [comment]
On my Mac, the `msg part` log message looks like:
```
11:04:11(write task): got msg part 7c6d0ace142c46978acc03dbed6de217
```
so your `got msg part ���{&` is definitely suspicious.

**Comment by stevengj** [comment]
@minrk, this is printing [the message "ident"](https://github.com/JuliaLang/IJulia.jl/blob/master/src/msg.jl#L70-L73) prefix, which is defined in the [Jupyter Messaging Documentation](https://jupyter-client.readthedocs.io/en/stable/messaging.html#the-wire-protocol) as `b'u-u-i-d',         # zmq identity(ies)`, described as `The front of the message is the ZeroMQ routing prefix, which can be zero or more socket identities.`

I always assumed that this was just an ASCII (or UTF-8) string of characters, but the documentation is very vague on this point.  Is it possible that it is UTF-16 or some other format on Windows?

**Comment by Datseris** [comment]
> so your `got msg part ���{&` is definitely suspicious.

I agree but I only thought that it was something like missing unicode characters (from example half the unicode stuff I use in Julia I can't see in the Terminal).

Oh something important: I am using Python version 3: a full anaconda installation for my conda stuff. Do you think that using python 2 or the default cond install would change things? I am trying now!!!

**Comment by Datseris** [comment]
Before moving conda to python 2, here is the change of making the protocol adaptare 5.3: (still freezes).
I also run it from the CMD instead of PowerShell:

```
C:\Users\datseris>jupyter notebook
[I 17:26:54.536 NotebookApp] Serving notebooks from local directory: C:\Users\datseris
[I 17:26:54.536 NotebookApp] 0 active kernels
[I 17:26:54.536 NotebookApp] The Jupyter Notebook is running at:
[I 17:26:54.536 NotebookApp] http://localhost:8888/?token=a5a668b8892fa1462468143c3b45f43dc949216dd614474d
[I 17:26:54.536 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 17:26:54.543 NotebookApp]

    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=a5a668b8892fa1462468143c3b45f43dc949216dd614474d
[I 17:26:54.799 NotebookApp] Accepting one-time-token-authenticated connection from ::1
[I 17:27:03.056 NotebookApp] Creating new notebook in
[I 17:27:04.249 NotebookApp] Kernel started: 9fdc8dcf-7757-4132-a622-0fb651f4dd96
┌ Warning: `srand` is deprecated, use `Random.seed!` instead.
│   caller = __init__() at init.jl:22
└ @ IJulia C:\Users\datseris\.julia\dev\IJulia\src\init.jl:22
PROFILE = Dict{String,Any}("key"=>"285c801b-791abb4de98a40bb26c7b75e","transport"=>"tcp","signature_scheme"=>"hmac-sha256","ip"=>"127.0.0.1","hb_port"=>53476,"control_port"=>53475,"shell_port"=>53472,"stdin_port"=>53474,"kernel_name"=>"julia-0.7","iopub_port"=>53473)
17:27:10(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict("username"=>"jlkernel","session"=>"UUID(\"6469d702-fa45-4b68-b9b5-374c80771675\")"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"da2643d4-73ad-42f4-8b01-4742d5c625b3\")","msg_type"=>"status","username"=>"jlkernel","date"=>2018-07-30T17:27:10.259,"version"=>"5.0","session"=>"UUID(\"6469d702-fa45-4b68-b9b5-374c80771675\")"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"starting")
}
Starting kernel event loops.
17:27:12(write task): got msg part ���
17:27:12(write task): got msg part <IDS|MSG>
17:27:12(write task): RECEIVED IPython Msg [ idents ���  ] {
  parent_header = Dict{String,Any}(),
  header = Dict{String,Any}("msg_id"=>"f01e736e-c6d3dd6ae806fdc037380d02","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-07-30T15:27:04.310000Z","version"=>"5.3","session"=>"997d2737c44842e883d2489c93ee97f3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}()
}
17:27:12(write task): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"f01e736e-c6d3dd6ae806fdc037380d02","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-07-30T15:27:04.310000Z","version"=>"5.3","session"=>"997d2737c44842e883d2489c93ee97f3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"42393c2d-1b22-432e-9d8b-1c4f636364f2\")","msg_type"=>"status","username"=>"username","date"=>2018-07-30T17:27:12.261,"version"=>"5.0","session"=>"997d2737c44842e883d2489c93ee97f3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"busy")
}
17:27:12(write task): SENDING IPython Msg [ idents ���  ] {
  parent_header = Dict{String,Any}("msg_id"=>"f01e736e-c6d3dd6ae806fdc037380d02","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-07-30T15:27:04.310000Z","version"=>"5.3","session"=>"997d2737c44842e883d2489c93ee97f3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"4d31ae84-6574-483d-9f6f-0a5b27e3b695\")","msg_type"=>"kernel_info_reply","username"=>"username","date"=>2018-07-30T17:27:12.494,"version"=>"5.0","session"=>"997d2737c44842e883d2489c93ee97f3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("help_links"=>Dict{String,String}[Dict("text"=>"Julia Home Page","url"=>"http://julialang.org/"), Dict("text"=>"Julia Documentation","url"=>"http://docs.julialang.org/"), Dict("text"=>"Julia Packages","url"=>"http://pkg.julialang.org/")],"language_info"=>Dict("file_extension"=>".jl","name"=>"julia","mimetype"=>"application/julia","version"=>"0.7.0"),"implementation"=>"ijulia","banner"=>"Julia: A fresh approach to technical computing.","protocol_version"=>"5.3")
}
17:27:12(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"f01e736e-c6d3dd6ae806fdc037380d02","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-07-30T15:27:04.310000Z","version"=>"5.3","session"=>"997d2737c44842e883d2489c93ee97f3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"13ca4e7b-2acf-456e-a427-c3688aeca4e5\")","msg_type"=>"status","username"=>"username","date"=>2018-07-30T17:27:12.794,"version"=>"5.0","session"=>"997d2737c44842e883d2489c93ee97f3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"idle")
}
```

**Comment by stevengj** [comment]
Can you try checking out the latest version of ZMQ?  (`dev ZMQ` at the package prompt).   I just fixed a memory bug there that may be relevant (JuliaInterop/ZMQ.jl#183).

**Comment by Datseris** [comment]
Nope, same as before... I still haven't moved to Python 2 though.

**Comment by stevengj** [comment]
You will need to do `ENV["IJULIA_DEBUG"]=true; Pkg.build("IJulia")` to get debugging output again.

**Comment by ChrisRackauckas** [comment]
I found an example the other day of a v0.7 nightly behavior where if code is compiled in a local scope the first time it can cause a compilation hang but if you do a few things first it won't hang.

https://github.com/JuliaLang/julia/issues/28356#issuecomment-409297092

I wonder if this is hitting something similar: debug modes makes it run code that somehow makes it not hang for the same unknown reason? It's a long shot but an idea.

**Comment by stevengj** [comment]
I don't think we have any functions with 396 arguments.  😉  Unfortunately, I don't spend much time with Windows, it's been difficult for me to debug this.

**Comment by ChrisRackauckas** [comment]
There is a second example there with no large argument functions.

**Comment by stevengj** [comment]
I would certainly be happy if a Julia update in the near future magically fixes this

**Comment by hessammehr** [comment]
@stevengj Interestingly, setting IJULIA_DEBUG as you suggested (see below) seems to fix the hang, but only for `jupyter console --kernel="julia-0.7"`; not `jupyter notebook` or `jupyter lab`.

```julia
ENV["IJULIA_DEBUG"] = true
(v0.7) pkg> build IJulia
```

**Comment by Datseris** [comment]
ASTONISHING NEWS EVERYONE!!!!

![image](https://user-images.githubusercontent.com/19669089/43639818-cb7c7832-971d-11e8-91af-f550cae3b991.png)

In current master and release candidate 1 it works!!!!! @stevengj you may close this if you want to.

**Comment by ChrisRackauckas** [comment]
I am getting hanging on v1.0.0-DEV still. I did create a separate issue for an example I could (almost) isolate on "running the same code in slightly different ways sometimes makes the compilation hang" https://github.com/JuliaLang/julia/issues/28417 . So maybe it's related. I've been finding a few of these pathological compilation bugs on Julia master and they are so weird but all with the same symptom so it's hard to know if they are actually the same.

**Comment by JeffBezanson** [comment]
Does this hang feature 100% cpu use, or is the process blocked?

**Comment by hessammehr** [comment]
On my computer the process is blocked in all cases.

**Comment by JeffBezanson** [comment]
If the process is blocked, then this is not similar to https://github.com/JuliaLang/julia/issues/28417. When the compiler runs off the rails it uses CPU.

**Comment by ChrisRackauckas** [comment]
Yeah it's not using CPU. So different issue, and I have no leads to check.

**Comment by hessammehr** [comment]
I've linked a [gist](https://gist.github.com/hessammehr/c0ea0ff0f1c791735f129726494edc27) showing the output of both `jupyter console` and `jupyter notebook`.  You can see that the `jupyter notebook` process gets stuck pretty early on waiting for Jupyter to respond. To me this seems more like a Jupyter than IJulia issue.

On observation is the `idents status`  message which looks a bit mangled in the case of `jupyter notebook`.

**Comment by hessammehr** [comment]
Even better if you [compare](https://gist.github.com/hessammehr/d1f8f46bf23f87418ad3e13983ce3c57) the output of `jupyter notebook --debug` for both Julia 0.6.4 and Julia 0.7rc2 (IJulia master, debug), you can see that Jupyter acknowledges receipt of `idents ...` only in the case of Julia 0.6.4. Could the culprit be slight differences in Dict serialization?

**Comment by sandeep-sandhu** [comment]
Same issue here with 0.7.0 release in windows.

The non-debug console messages from running Jupyter are:
```

[I 11:04:49.669 NotebookApp] Kernel started: b43e61cd-7e90-4ee1-9850-d0d19401a43c
[ Info: Recompiling stale cache file C:\Users\myhome\.julia\compiled\v0.7\IJulia\nfu7T.ji for IJulia [7073ff75-c697-5162-941a-fcdaad2a7d2a]
┌ Warning: `srand` is deprecated, use `Random.seed!` instead.
│   caller = top-level scope at none:0
└ @ Core none:0
┌ Warning: `srand` is deprecated, use `Random.seed!` instead.
│   caller = __init__() at init.jl:25
└ @ IJulia C:\Users\myhome\.julia\packages\IJulia\RBfPg\src\init.jl:25
[W 11:04:59.870 NotebookApp] Timeout waiting for kernel_info reply from b43e61cd-7e90-4ee1-9850-d0d19401a43c
Starting kernel event loops.
[I 11:05:03.576 NotebookApp] Adapting to protocol v5.0 for kernel b43e61cd-7e90-4ee1-9850-d0d19401a43c
```

**Comment by notZaki** [comment]
It worked for me when I enabled the debugging, but only on the first notebook attempt. 
After a bit of trial and error, here's a setup that kinda works for me (on julia 0.7):

- in the Julia REPL, follow the [previous instruction to enable debugging](https://github.com/JuliaLang/IJulia.jl/issues/693#issuecomment-408707684):
```
ENV["IJULIA_DEBUG"]=true
] dev IJulia
] build IJulia
```
- in powershell, launch jupyter using either `jupyter notebook` or `jupyter lab` if installed
    + I'm using `jupyter lab` from an existing Anaconda installation (Python3)
- create a new notebook using the 0.7 kernel, or open existing one, and quickly type any simple command and execute it, e.g. `1+1` and shift+enter. 
   + In the background, the powershell window will show something along the lines of `[ Info: Precompiling IJulia...]` (or recompiling stale cache) when the notebook is first opened. After a couple of seconds, it'll then say `PROFILE = Dict{String,Any}...` followed by a deluge of lines. The goal is to execute a command in the notebook **before** `PROFILE...` and his buddies show up--that's why a command has to be entered into the notebook quickly after opening it.

If successful, then all future commands in that specific notebook will work.
Opening another new/existing notebook will make that notebook stall. 
In other words, the first notebook that you open/create will likely be the only one that'll work during the jupyter session.

If jupyter is closed, then relaunching it from powershell doesn't work--all the notebooks stall. The workaround is to delete the compiled version of IJulia, which is typically located in `C:\Users\username\.julia\compiled\v0.7\IJulia`. I just delete the entire `IJulia` folder before launching jupyter. 
After deleting the precompiled version, it's the same deal as before: open notebook and quickly execute a command in a notebook while IJulia is precompiling, and then that specific notebook will remain usable.

This is by no means a proper solution, but it may be helpful for figuring out the cause? Assuming anyone else can reproduce the steps to get a single notebook working.
It is weird that the problems only starts to occur after precompiling... and the entire thing only works if debugging is enabled. 😕

**Comment by hhaensel** [comment]
Still no clue, how I could start debugging :-(
I realised, though, that when starting from the REPL, the kernel icon is changing to "ready" at some point, whereas it does not, when I launch jupyter from the powershell.

**Comment by chromezh** [comment]
Hi everyone. I am using Windows + Julia 0.7. I encountered a similar error, but **my error occurred when installing the package**:

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0 (2018-08-08 06:46 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-w64-mingw32

julia> Pkg.add("IJulia")
WARNING: Base.Pkg is deprecated, run `using Pkg` instead
 in module Main
   Cloning default registries into E:\julia-depot\registries
   Cloning registry General from "https://github.com/JuliaRegistries/General.git"
  Updating registry at `E:\julia-depot\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
 Resolving package versions...
 Installed VersionParsing ─ v1.1.1
 Installed Conda ────────── v1.0.1
 Installed ZMQ ──────────── v0.6.3
 Installed BinaryProvider ─ v0.4.0
 Installed IJulia ───────── v1.9.2
 Installed JSON ─────────── v0.19.0
 Installed MbedTLS ──────── v0.6.0
 Installed Compat ───────── v1.0.1
  Updating `E:\julia-depot\environments\v0.7\Project.toml`
  [7073ff75] + IJulia v1.9.2
  Updating `E:\julia-depot\environments\v0.7\Manifest.toml`
  [b99e7846] + BinaryProvider v0.4.0
  [34da2185] + Compat v1.0.1
  [8f4d0f93] + Conda v1.0.1
  [7073ff75] + IJulia v1.9.2
  [682c06a0] + JSON v0.19.0
  [739be429] + MbedTLS v0.6.0
  [81def892] + VersionParsing v1.1.1
  [c2297ded] + ZMQ v0.6.3
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8bb1440f] + DelimitedFiles
  [8ba89e20] + Distributed
  [7b1f6079] + FileWatching
  [b77e0a4c] + InteractiveUtils
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  Building Conda ──→ `E:\julia-depot\packages\Conda\m7vem\deps\build.log`
  Building MbedTLS → `E:\julia-depot\packages\MbedTLS\Qo8TN\deps\build.log`
┌ Error: Error building `MbedTLS`:
│ [ Info: Downloading https://github.com/JuliaWeb/MbedTLSBuilder/releases/download/v0.11/MbedTLS.v1.0.0.x86_64-w64-mingw32.tar.gz to E:\julia-depot\packages\MbedTLS\Qo8TN\deps\usr\downloads\MbedTLS.v1.0.0.x86_64-w64-mingw32.tar.gz...
│ ERROR: LoadError: Could not download https://github.com/JuliaWeb/MbedTLSBuilder/releases/download/v0.11/MbedTLS.v1.0.0.x86_64-w64-mingw32.tar.gz to E:\julia-depot\packages\MbedTLS\Qo8TN\deps\usr\downloads\MbedTLS.v1.0.0.x86_64-w64-mingw32.tar.gz
│ Stacktrace:
│  [1] error(::String) at .\error.jl:33
│  [2] macro expansion at .\logging.jl:313 [inlined]
│  [3] #download#93(::Bool, ::Function, ::String, ::String) at E:\julia-depot\packages\BinaryProvider\ZbFxn\src\PlatformEngines.jl:443
│  [4] #download at .\none:0 [inlined]
│  [5] #download_verify#94(::Bool, ::Bool, ::Bool, ::Function, ::String, ::String, ::String) at E:\julia-depot\packages\BinaryProvider\ZbFxn\src\PlatformEngines.jl:523
│  [6] #download_verify at .\none:0 [inlined]
│  [7] #install#115(::Prefix, ::String, ::Bool, ::Bool, ::Bool, ::Function, ::String, ::String) at E:\julia-depot\packages\BinaryProvider\ZbFxn\src\Prefix.jl:348
│  [8] (::getfield(BinaryProvider, Symbol("#kw##install")))(::NamedTuple{(:prefix, :force, :verbose),Tuple{Prefix,Bool,Bool}}, ::typeof(install), ::String, ::String) at .\none:0
│  [9] top-level scope at E:\julia-depot\packages\MbedTLS\Qo8TN\deps\build.jl:45
│  [10] include at .\boot.jl:317 [inlined]
│  [11] include_relative(::Module, ::String) at .\loading.jl:1038
│  [12] include(::Module, ::String) at .\sysimg.jl:29
│  [13] include(::String) at .\client.jl:398
│  [14] top-level scope at none:0
│ in expression starting at E:\julia-depot\packages\MbedTLS\Qo8TN\deps\build.jl:41
│ Exception calling "DownloadFile" with "2" argument(s): "Unable to connect to the remote server"
│ At line:5 char:1
│ + $webclient.DownloadFile("https://github.com/JuliaWeb/MbedTLSBuilder/r ...
│ + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
│     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
│     + FullyQualifiedErrorId : WebException
│
└ @ Pkg.Operations C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\Pkg\src\Operations.jl:1068
  Building ZMQ ────→ `E:\julia-depot\packages\ZMQ\GA4Xh\deps\build.log`
  Building IJulia ─→ `E:\julia-depot\packages\IJulia\9RcVi\deps\build.log`

julia> using IJulia
[ Info: Precompiling IJulia [7073ff75-c697-5162-941a-fcdaad2a7d2a]
ERROR: LoadError: MbedTLS not installed properly, run Pkg.build("MbedTLS"), restart Julia and try again
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] top-level scope at E:\julia-depot\packages\MbedTLS\Qo8TN\src\MbedTLS.jl:53
 [3] include at .\boot.jl:317 [inlined]
 [4] include_relative(::Module, ::String) at .\loading.jl:1038
 [5] include(::Module, ::String) at .\sysimg.jl:29
 [6] top-level scope at none:2
 [7] eval at .\boot.jl:319 [inlined]
 [8] eval(::Expr) at .\client.jl:399
 [9] top-level scope at .\none:3
in expression starting at E:\julia-depot\packages\MbedTLS\Qo8TN\src\MbedTLS.jl:52
ERROR: LoadError: LoadError: Failed to precompile MbedTLS [739be429-bea8-5141-9913-cc70e7f3736d] to E:\julia-depot\compiled\v0.7\MbedTLS\bf9T0.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] macro expansion at .\logging.jl:313 [inlined]
 [3] compilecache(::Base.PkgId, ::String) at .\loading.jl:1185
 [4] _require(::Base.PkgId) at .\logging.jl:311
 [5] require(::Base.PkgId) at .\loading.jl:852
 [6] macro expansion at .\logging.jl:311 [inlined]
 [7] require(::Module, ::Symbol) at .\loading.jl:834
 [8] include at .\boot.jl:317 [inlined]
 [9] include_relative(::Module, ::String) at .\loading.jl:1038
 [10] include at .\sysimg.jl:29 [inlined]
 [11] include(::String) at E:\julia-depot\packages\IJulia\9RcVi\src\IJulia.jl:35
 [12] top-level scope at none:0
 [13] include at .\boot.jl:317 [inlined]
 [14] include_relative(::Module, ::String) at .\loading.jl:1038
 [15] include(::Module, ::String) at .\sysimg.jl:29
 [16] top-level scope at none:2
 [17] eval at .\boot.jl:319 [inlined]
 [18] eval(::Expr) at .\client.jl:399
 [19] top-level scope at .\none:3
in expression starting at E:\julia-depot\packages\IJulia\9RcVi\src\hmac.jl:1
in expression starting at E:\julia-depot\packages\IJulia\9RcVi\src\IJulia.jl:355
ERROR: Failed to precompile IJulia [7073ff75-c697-5162-941a-fcdaad2a7d2a] to E:\julia-depot\compiled\v0.7\IJulia\nfu7T.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] macro expansion at .\logging.jl:313 [inlined]
 [3] compilecache(::Base.PkgId, ::String) at .\loading.jl:1185
 [4] _require(::Base.PkgId) at .\logging.jl:311
 [5] require(::Base.PkgId) at .\loading.jl:852
 [6] macro expansion at .\logging.jl:311 [inlined]
 [7] require(::Module, ::Symbol) at .\loading.jl:834

julia> notebook()
ERROR: UndefVarError: notebook not defined
Stacktrace:
 [1] top-level scope at none:0

julia>
```

**Comment by chromezh** [comment]
Sorry for my noise. It is just because of the web connection.

**Comment by patdub** [comment]
Hello,

I have the same problem since updating from Julia 0.6.4 to 0.7 and also 1.0.

No error message from the shell but going to the Help menu then about, the message windows says that no kernel is available but the bar message says connected then trusted.

![image](https://user-images.githubusercontent.com/42249778/44031134-50bd9eb2-9f03-11e8-82a0-8fc9e66b72c3.png)

**Comment by d4n1elchen** [comment]
Hello,

As for the weird characters in idents, I found that it only shows in the first `kernel_info_request` on my computer. The idents of following msgs are looking good. I'm using Julia 1.0.0.

```
[I 08:59:21.255 NotebookApp] Kernel restarted: 2e08217d-b5e5-4ee0-8811-b33e76b7a6a8
[ Info: Recompiling stale cache file C:\Users\daniel\.julia\compiled\v1.0\IJulia\nfu7T.ji for IJulia [7073ff75-c697-5162-941a-fcdaad2a7d2a]
[I 08:59:22.217 NotebookApp] Saving file at /julia-learn/Untitled.ipynb
PROFILE = Dict{String,Any}("key"=>"a48b10b8-9149fb2c76f08aedf414bc4d","transport"=>"tcp","signature_scheme"=>"hmac-sha256","shell_port"=>65186,"hb_port"=>65190,"control_port"=>65189,"ip"=>"127.0.0.1","stdin_port"=>65188,"iopub_port"=>65187,"kernel_name"=>"julia-1.0")
08:59:26(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict("username"=>"jlkernel","session"=>"UUID(\"a3e77d51-2b77-40e2-a6d3-1c13ade3937e\")"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"4efff4c5-c00f-4f0e-8dd3-dc54759b6875\")","msg_type"=>"status","username"=>"jlkernel","date"=>2018-08-14T08:59:26.028,"version"=>"5.0","session"=>"UUID(\"a3e77d51-2b77-40e2-a6d3-1c13ade3937e\")"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"starting")
}
Starting kernel event loops.
08:59:27(write task): got msg part ���t
08:59:27(write task): got msg part <IDS|MSG>
08:59:27(write task): RECEIVED IPython Msg [ idents ���t ] {
  parent_header = Dict{String,Any}(),
  header = Dict{String,Any}("msg_id"=>"65971804-f94b7258b5c9d2aad0c07b2c","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:21.272207Z","version"=>"5.3","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}()
}
08:59:27(write task): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"65971804-f94b7258b5c9d2aad0c07b2c","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:21.272207Z","version"=>"5.3","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"7c206f54-bc71-4242-86c4-87c739f520fb\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.412,"version"=>"5.0","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"busy")
}
08:59:27(write task): SENDING IPython Msg [ idents ���t ] {
  parent_header = Dict{String,Any}("msg_id"=>"65971804-f94b7258b5c9d2aad0c07b2c","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:21.272207Z","version"=>"5.3","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"c325a8a0-f643-4bf4-9fb2-6858c7f1282b\")","msg_type"=>"kernel_info_reply","username"=>"username","date"=>2018-08-14T08:59:27.538,"version"=>"5.0","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("help_links"=>Dict{String,String}[Dict("text"=>"Julia Home Page","url"=>"http://julialang.org/"), Dict("text"=>"Julia Documentation","url"=>"http://docs.julialang.org/"), Dict("text"=>"Julia Packages","url"=>"http://pkg.julialang.org/")],"language_info"=>Dict("file_extension"=>".jl","name"=>"julia","mimetype"=>"application/julia","version"=>"1.0.0"),"implementation"=>"ijulia","banner"=>"Julia: A fresh approach to technical computing.","protocol_version"=>"5.0")
}
[I 08:59:27.693 NotebookApp] Adapting to protocol v5.0 for kernel 2e08217d-b5e5-4ee0-8811-b33e76b7a6a8
[I 08:59:27.695 NotebookApp] Restoring connection for 2e08217d-b5e5-4ee0-8811-b33e76b7a6a8:cf8677b67a1b43de98b7d61d028ceaa3
[I 08:59:27.696 NotebookApp] Replaying 3 buffered messages
08:59:27(write task): got msg part cf8677b67a1b43de98b7d61d028ceaa3
08:59:27(write task): got msg part <IDS|MSG>
08:59:27(write task): RECEIVED IPython Msg [ idents cf8677b67a1b43de98b7d61d028ceaa3 ] {
  parent_header = Dict{String,Any}(),
  header = Dict{String,Any}("msg_id"=>"094f1123340c4920848f9193ea775d51","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.701012Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}()
}
08:59:27(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"65971804-f94b7258b5c9d2aad0c07b2c","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:21.272207Z","version"=>"5.3","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"2b987c9e-89b9-4ae7-8536-ac6b7abb0687\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.733,"version"=>"5.0","session"=>"f9480d6a-cc3ee08d1526a59ab983eed4"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"idle")
}
08:59:27(write task): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"094f1123340c4920848f9193ea775d51","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.701012Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"ddd72462-7d4c-4350-a79b-536b8f4697b4\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.734,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"busy")
}
08:59:27(write task): SENDING IPython Msg [ idents cf8677b67a1b43de98b7d61d028ceaa3 ] {
  parent_header = Dict{String,Any}("msg_id"=>"094f1123340c4920848f9193ea775d51","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.701012Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"abedf074-43ca-442d-9ab8-51ed95719c1d\")","msg_type"=>"kernel_info_reply","username"=>"username","date"=>2018-08-14T08:59:27.737,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("help_links"=>Dict{String,String}[Dict("text"=>"Julia Home Page","url"=>"http://julialang.org/"), Dict("text"=>"Julia Documentation","url"=>"http://docs.julialang.org/"), Dict("text"=>"Julia Packages","url"=>"http://pkg.julialang.org/")],"language_info"=>Dict("file_extension"=>".jl","name"=>"julia","mimetype"=>"application/julia","version"=>"1.0.0"),"implementation"=>"ijulia","banner"=>"Julia: A fresh approach to technical computing.","protocol_version"=>"5.0")
}
08:59:27(write task): got msg part cf8677b67a1b43de98b7d61d028ceaa3
08:59:27(write task): got msg part <IDS|MSG>
08:59:27(write task): RECEIVED IPython Msg [ idents cf8677b67a1b43de98b7d61d028ceaa3 ] {
  parent_header = Dict{String,Any}(),
  header = Dict{String,Any}("msg_id"=>"ff0f50358ee949868f03808415250586","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.702009Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}()
}
08:59:27(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"094f1123340c4920848f9193ea775d51","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.701012Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"852828cf-6979-423a-b83b-4bfe888446e7\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.739,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"idle")
}
08:59:27(write task): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"ff0f50358ee949868f03808415250586","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.702009Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"664a147a-4525-4673-a2d1-9d427b4ef36a\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.74,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"busy")
}
08:59:27(write task): SENDING IPython Msg [ idents cf8677b67a1b43de98b7d61d028ceaa3 ] {
  parent_header = Dict{String,Any}("msg_id"=>"ff0f50358ee949868f03808415250586","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.702009Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"d2b90385-89d3-4454-bd05-5267025fffa5\")","msg_type"=>"kernel_info_reply","username"=>"username","date"=>2018-08-14T08:59:27.748,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("help_links"=>Dict{String,String}[Dict("text"=>"Julia Home Page","url"=>"http://julialang.org/"), Dict("text"=>"Julia Documentation","url"=>"http://docs.julialang.org/"), Dict("text"=>"Julia Packages","url"=>"http://pkg.julialang.org/")],"language_info"=>Dict("file_extension"=>".jl","name"=>"julia","mimetype"=>"application/julia","version"=>"1.0.0"),"implementation"=>"ijulia","banner"=>"Julia: A fresh approach to technical computing.","protocol_version"=>"5.0")
}
08:59:27(): SENDING IPython Msg [ idents status ] {
  parent_header = Dict{String,Any}("msg_id"=>"ff0f50358ee949868f03808415250586","msg_type"=>"kernel_info_request","username"=>"username","date"=>"2018-08-14T00:59:27.702009Z","version"=>"5.2","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"fa824f23-4afe-45cf-8e31-dbb8cfdeb240\")","msg_type"=>"status","username"=>"username","date"=>2018-08-14T08:59:27.75,"version"=>"5.0","session"=>"cf8677b67a1b43de98b7d61d028ceaa3"),
  metadata = Dict{String,Any}(),
  content = Dict("execution_state"=>"idle")
}
```

**Comment by c123w** [comment]
I'll just copy and paste from #702 what I found that (sort of) worked for Julia 1.0:

I've done a bit of experimenting, I managed to get IJulia to work by running this in the Julia REPL:
```
(v1.0) pkg> add IJulia
ENV["IJULIA_DEBUG"] = true
(v1.0) pkg> dev IJulia
(v1.0) pkg> build IJulia
```
 followed by `jupyter console --kernel="julia-1.0"` from powershell (not `jupyter-lab` or `jupyter-notebook` from powershell, or `notebook()` from the Julia REPL).

Doing only either one of these two things gave me the same hanging kernel as before.

**Comment by d4n1elchen** [comment]
@c123w 's solution doesn't work for me.

I think I got ways to solve this problem when I was trying to debug it but I don't know why it works.

First, I turn on the debugging output by rebuilding the package with `IJULIA_DEBUG = true`. Then I found there are some printing functions in the source code so I try to use them to print messages to the terminal. When I added the following code to `msg.jl` ... things get working!!
```diff
...
function recv_ipython(socket)
    lock(socket_locks[socket])
    try
+        @vprintln("")
        msg = recv(socket)
...
```
Not sure why it is but I think it should be related to the buffer of stdout.

**Comment by adrpo** [comment]
@d4n1elchen: thanks! your solution was the only one that worked for me. 
Windows 10, Julia 1.0.0, jupyter-notebook version 5.5.0 (Python 2.7.15 :: Anaconda, Inc).

**Comment by mykelk** [comment]
@d4n1elchen that worked for me as well on Windows 10, Julia 1.0.0, Jupyter 5.6.0, Python 2.7.15 (anaconda), all 64 bit

**Comment by d4n1elchen** [comment]
Here's some note about my previous solution. That solution only works if you turn on the `IJULIA_DEBUG`, which I think it does not truly solve the problem.

I think this problem should be related to #491 and #563. Previously there's an issue about when to send the idle message. At that time, the solution was adding a delay before sending the idle message. I think it's because of the delay, the code, together with `flush_all()`, was wrapped by `@asnyc`. That means the `flush_all()` won't be called right after message handler finish the jobs. I think that will cause unexpected flush for the following messages. So I move the `flush_all()` out from the `@asnyc` wrapping and remain everything being the same (also remove `@vprintln` I just added). Then everything looks fine. It works with IJULIA_DEBUG is either true or false.

And according to #563, the delay is no more needed and has been removed. I think the `@async` is also not needed. But I'm not sure about this so I keep it in the code.

I'll make a pull request about the changes above.

**Comment by stevengj** [comment]
My suspicion is that this is some subtle memory bug in ZMQ.jl (or a julia compiler bug), which is why it is appearing and disappearing randomly when you do unrelated things (move a flush statement, turn on debugging output, etcetera).

**Comment by dierickx3** [comment]
Just doing the following that somebody posted above from REPL seems to have fixed it for me. Like others, I was getting the hang no matter what I entered, which led to this thread from a google search. The below _seems_ to have fixed it for me for now.

> (v1.0) pkg> add IJulia
> julia> ENV["IJULIA_DEBUG"] = true
> (v1.0) pkg> dev IJulia
> (v1.0) pkg> build IJulia

_note: Windows 10, Julia 1.0, Anaconda, Python 3.6.3, Jupyter 5.6, Firefox 61:_

![image](https://user-images.githubusercontent.com/4108909/44302842-a2b18e80-a2f6-11e8-9ead-fb50473629ac.png)

**Comment by strickek** [comment]
For me it worke now with master, only additional 
``
(v1.0) pkg> dev IJulia
``
(no build, no ENV... necessary.

**Comment by wanyancan** [comment]
I tried to use dev version. 
With 1+1, the return status can be correctly generated. 
but with input "?println"
it will generate output to 'stdout', 
but there's never a response changing status back to "idle"
So the block still hangs forever.

> 21:03:59(write task): [EXECUTING D 21:03:59.059 NotebookApp] activity on 7c8563fd-7376-4f1d-a599-83ad692592f4: status
?println
21:03:59(write task): SENDING IPython Msg [ idents execute_input ] {
  parent_header = Dict{String,Any}("msg_id"=>"f7e8a31e6ed44cffb643089d5afbe303","msg_type"=>"execute_request","username"=>"username","date"=>"2018-08-19T13:03:59.053000Z","version"=>"5.2","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"475d86bc-0951-4e59-9d76-640b8786b273\")","msg_type"=>"execute_input","username"=>"username","date"=>2018-08-19T21:03:59.062,"version"=>"5.0","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  metadata = Dict{String,Any}(),
  content = Dict{String,Any}("code"=>"?println","execution_count"=>2)
}
[D 21:03:59.071 NotebookApp] activity on 7c8563fd-7376-4f1d-a599-83ad692592f4: execute_input
21:03:59(send stdout task): SENDING IPython Msg [ idents stdout ] {
  parent_header = Dict{String,Any}("msg_id"=>"f7e8a31e6ed44cffb643089d5afbe303","msg_type"=>"execute_request","username"=>"username","date"=>"2018-08-19T13:03:59.053000Z","version"=>"5.2","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"bb9e969b-309e-4afc-bfb3-d37e396fc982\")","msg_type"=>"stream","username"=>"username","date"=>2018-08-19T21:03:59.408,"version"=>"5.0","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  metadata = Dict{String,Any}(),
  content = Dict("name"=>"stdout","text"=>"search:")
}
[D 21:03:59.493 NotebookApp] activity on 7c8563fd-7376-4f1d-a599-83ad692592f4: stream
21:03:59(send stdout task): SENDING IPython Msg [ idents stdout ] {
  parent_header = Dict{String,Any}("msg_id"=>"f7e8a31e6ed44cffb643089d5afbe303","msg_type"=>"execute_request","username"=>"username","date"=>"2018-08-19T13:03:59.053000Z","version"=>"5.2","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  header = Dict{String,Any}("msg_id"=>"UUID(\"65f543e5-a88e-4b49-b8f2-733f2ff4a105\")","msg_type"=>"stream","username"=>"username","date"=>2018-08-19T21:03:59.514,"version"=>"5.0","session"=>"eba20455eaa54116aedd7e3a59bd7600"),
  metadata = Dict{String,Any}(),
  content = Dict("name"=>"stdout","text"=>" \e[0m\e[1mp\e[22m\e[0m\e[1mr\e[22m\e[0m\e[1mi\e[22m\e[0m\e[1mn\e[22m\e[0m\e[1mt\e[22m\e[0m\e[1ml\e[22m\e[0m\e[1mn\e[22m")
}
[D 21:03:59.520 NotebookApp] activity on 7c8563fd-7376-4f1d-a599-83ad692592f4: stream
[D 21:04:42.216 NotebookApp] 200 GET /api/contents/JuliaBoxTutorials/intro-to-julia/00.%20Jupyter_notebooks.ipynb?content=0&_=1534683761575 (::1) 3.00ms
[I 21:04:42.224 NotebookApp] Saving file at /JuliaBoxTutorials/intro-to-julia/00. Jupyter_notebooks.ipynb
[D 21:04:42.226 NotebookApp] Saving C:\julia_projs\JuliaBoxTutorials\intro-to-julia\00. Jupyter_notebooks.ipynb
[D 21:04:42.292 NotebookApp] 200 PUT /api/contents/JuliaBoxTutorials/intro-to-julia/00.%20Jupyter_notebooks.ipynb (::1) 68.00ms

**Comment by Bob-McCrory** [comment]
Just installed Julia/Jupyter/IJulia on a Surface Pro 2017 w/ an i7 processor. IJulia is either hanging randomly - I can detect no pattern - or it is running so incredibly slow that it appears to hang. I'm using the development IJulia as suggested above. As it stands, IJulia is unusable.

**Comment by korsbo** [comment]
I think that I have made some progress on this issue. I suspect that there may be some flushing error with `send`. I have created a very hacky fix for it, but since it may be good if someone actually fixed the deeper problem I will report how I pinpointed the bug.


### The hacky fix

In the file msg.jl, add the line 

```Julia
print(orig_stdout[], "")
```
to the end of the `send_ipython` function (I placed it in the `finally` block at line 62). It must be placed below all the `send` calls, but I have not found any other restrictions.


It would be good if someone could confirm that this is working for them as well.

### Elaboration on the error

Just in case it is helpful to the devs.

I got the verbose debug printouts to work by just manually changing `const IJULIA_DEBUG = false` to `true` in `deps/deps.jl`. 

The behaviour of IJulia was not really consistent. Sometimes everything worked, sometimes it did not. 
To find this error, I looked at cases where I could execute a cell at least once, but where it then stopped working. In those cases, the following happens:

1. `execute_requests` gets called.
2. Everything works consistently fine until one reaches 
```Julia
        send_ipython(requests[],
                     msg_reply(msg, "execute_reply",
                               Dict("status" => "ok",
                                            "payload" => execute_payloads,
                                            "execution_count" => n,
                                            "user_expressions" => user_expressions)))
```
at line 228 of execute_request.jl.

3. `send_ipython` does evaluate, but the next step of the process gets blocked (I'm not sure where, I only have a hack-fix for it). 

When I added printout messages to the end of `send_ipython` for debugging, everything suddenly worked fine. 

I think that there are more calls to `send_ipython` which cause errors. Sometimes one cannot do a first evaluation, and although the `send_ipython` ugly fix still helps, the trace of how one got to that function may differ. It is however notable that the two other calls to `send_ipython` from within `execute_request` did not freeze (I repeated this many times). 

I suspect that this is some kind of flushing error and that it may be `send`s fault. However, I am not certain of this. If I am right, then the real fix should probably be implemented there rather than adding an empty printout to IJulia.

Anyway, I hope this helps. Life without IJulia has been tough and I am now much happier than I was an hour ago.

**Comment by korsbo** [comment]
The plot thickens. My ugly hack-fix seems to only reduce the likelihood of snags. Also the likelihood of getting stuck is different for different commands. By holding Ctrl-Enter one can see roughly how long it takes before it will lock jupyter.

Low likelihood of snags:
```julia
1+1 # usually takes on the order of 1000 evals for me.
eval( :(1 + 2) ) # similar to just 1+1 
using Pkg
print("hello") # takes <100 evals
@macroexpand @time 1/2
```

Snags after just a few evaluations:
```julia
eval( :(print("hello") )
@time 1/2
```

Stuff that I have not seen working even once:
```julia
pkg.status()
```

I don't know what sourcery this is, but since not everyone can reproduce these errors I figured that I'd share the info regardless.

**Comment by XinyuuLi** [comment]
Same problem with Julia 1.0.0 + win10. All solutions in this issue didn't work.

**Comment by ceysa75** [comment]
Same here. For days now struggeling with that issue. I wonder if there is any statement by the developers in this regard. Or is it just a minority of the users experiencing this ? Worth waiting or should I make my first steps with Julia 0.6x ?

**Comment by ChrisRackauckas** [comment]
I think the underlying issue is that the majority of Julia developers are on Mac and Linux which is why this took some time to get noticed. But also, it's just a really hard issue to reproduce so it will take some time to work through. The best way to help right now is probably to find minimal reproducible examples of this working/not working to attempt to narrow down what's wrong to help find a fix. Until it is fixed, I would recommend using the Julia REPL or an IDE (VScode, Juno, etc.) if on Windows and Julia v1.0, or reverting to Julia v0.6.x (on Windows. Mac and Linux seem to be fine given the bug reports).

**Comment by hessammehr** [comment]
For folks stuck with this bug, here is a workaround that seems to consistently work for me (Julia 1.0, IJulia master, Jupyter lab 0.34.1, Windows 10): Simply switch the kernel to Julia 0.6.4, maybe evaluate something simple like `1` for good measure, then switch back to Julia 1.0.

Ugly but hope this helps anyone on Julia 1.0 who needs to get things done.

**Comment by Nectarineimp** [comment]
I just did an update of IJulia and I notice it is working for 1.0 on Windows 10! Looks good!

**Comment by Bob-McCrory** [comment]
[LScratch00.zip](https://github.com/JuliaLang/IJulia.jl/files/2329201/LScratch00.zip)
Sorry, it still locks up for me. I've attached a zip of a fairly simple notebook - though it has graphs - that IJulia keeps locking up on.

Hope this helps.

**Comment by stevengj** [comment]
These random workarounds that people keep posting are not particularly helpful.

The error must be some kind of memory bug or race condition, which is why it is hard to reproduce and why doing seemingly things may make it go away temporarily.   Further descriptions of random events don't help to track down the problem.

**Comment by jlperla** [comment]
Likewise, still doesn't work (and the workarounds don't either).  Trivial notebook still locks on Windows.  For the versions I had just done
```julia
(v1.0) pkg> up IJulia
  Updating registry at `C:\Users\jlperla\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
 Resolving package versions...
 Installed SoftGlobalScope ─ v1.0.0
 Installed IJulia ────────── v1.10.0
 Installed Revise ────────── v0.7.3
  Updating `C:\Users\jlperla\.julia\environments\v1.0\Project.toml`
  [7073ff75] ↑ IJulia v1.9.3 ⇒ v1.10.0
  [295af30f] ↑ Revise v0.7.2 ⇒ v0.7.3
  Updating `C:\Users\jlperla\.julia\environments\v1.0\Manifest.toml`
  [7073ff75] ↑ IJulia v1.9.3 ⇒ v1.10.0
  [295af30f] ↑ Revise v0.7.2 ⇒ v0.7.3
  [b85f4697] + SoftGlobalScope v1.0.0
  Building IJulia → `C:\Users\jlperla\.julia\packages\IJulia\fDY4W\deps\build.log`

(v1.0) pkg> precompile
Precompiling project...
Precompiling Revise
[ Info: Recompiling stale cache file C:\Users\jlperla\.julia\compiled\v1.0\Revise\M1Qoh.ji for Revise [295af30f-e4ad-537b-8983-00126c2a3abe]
Precompiling IJulia
[ Info: Recompiling stale cache file C:\Users\jlperla\.julia\compiled\v1.0\IJulia\nfu7T.ji for IJulia [7073ff75-c697-5162-941a-fcdaad2a7d2a]
```

**Comment by finmod** [comment]
Let me tell you how I fixed this on Windows 10.

1. Scarp the Julia v1.0 recent install by deleting .julia folder
2. Update Anaconda3 or Miniconda to latest in particular Jupyter
3. Delete redundant Jupyter kernels in $HOME\AppData\Roaming\jupyter\kernels\
4. Install Julia v1.0
5. set ENV["JUPYTER"] = "C:\\Users............\\Anaconda3\\Scripts\\jupyter"
pkg> add IJulia 

This is it.

**Comment by wanyancan** [comment]
@finmod No. It won't solve the issue. IJulia still hangs up randomly.

Do not post trivial fix unless you have found the root of the problem.

**Comment by filmor** [comment]
Oddly enough, the broken identifier seems to always be followed by an unbroken `"<IDS|MSG>"` tag, I have until now only seen issues on the initial `recv`.

**Comment by ChrisRackauckas** [comment]
I get random hangs even with debug mode on. I did get a kernel error though `'_xsrf' argument missing from POST`. Maybe that's a lead.

**Comment by stevengj** [comment]
In the last week, I've been going through ZMQ.jl line-by-line and overhauling it, since it dates back to Julia 0.1 and does a lot of low-level things that needed re-examining.  I found several more places in the code that were unsafe about rooting Julia objects, which might run into trouble with the more aggressive garbage collection in 1.0 compared to 0.6.

Once JuliaInterop/ZMQ.jl#186 is merged, I will make a branch of IJulia that uses the new ZMQ API, and see if that helps on Windows.

**Comment by stevengj** [comment]
See JuliaLang/IJulia.jl#729 … maybe someone on Windows could check out ZMQ `master` and the IJulia `newzmq` branch and give it a spin.

**Comment by XinyuuLi** [comment]
@stevengj Update ZMQ makes it work for me. Thanks for your work. 
```
julia> Pkg.add("ZMQ")
 Resolving package versions...
 Installed SoftGlobalScope ─ v1.0.2
 Installed IJulia ────────── v1.10.0
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Project.toml`
  [7073ff75] ↑ IJulia v1.9.3 ⇒ v1.10.0
  [c2297ded] + ZMQ v0.6.4
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Manifest.toml`
  [7073ff75] ↑ IJulia v1.9.3 ⇒ v1.10.0
  [b85f4697] + SoftGlobalScope v1.0.2
  Building IJulia → `C:\Users\Xinyu\.julia\packages\IJulia\fDY4W\deps\build.log`
```

**Comment by stevengj** [comment]
@XinyuuLi, to get the latest version you need to `add ZMQ#master` and `add IJulia#newzmq` at the Pkg prompt.

**Comment by XinyuuLi** [comment]
@stevengj 
```
(v1.0) pkg> add ZMQ#master
  Updating registry at `C:\Users\Xinyu\.julia\registries\General`
  Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Cloning git-repo `https://github.com/JuliaInterop/ZMQ.jl.git`
  Updating git-repo `https://github.com/JuliaInterop/ZMQ.jl.git`
 Resolving package versions...
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Project.toml`
  [c2297ded] ↑ ZMQ v0.6.4 ⇒ v0.6.4+ #master (https://github.com/JuliaInterop/ZMQ.jl.git)
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Manifest.toml`
  [c2297ded] ↑ ZMQ v0.6.4 ⇒ v0.6.4+ #master (https://github.com/JuliaInterop/ZMQ.jl.git)
  Building ZMQ → `C:\Users\Xinyu\.julia\packages\ZMQ\ABGOx\deps\build.log`

(v1.0) pkg> add IJulia#newzmq
   Cloning git-repo `https://github.com/JuliaLang/IJulia.jl.git`
  Updating git-repo `https://github.com/JuliaLang/IJulia.jl.git`
 Resolving package versions...
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Project.toml`
  [7073ff75] ↑ IJulia v1.10.0 ⇒ v1.10.0+ #newzmq (https://github.com/JuliaLang/IJulia.jl.git)
  Updating `C:\Users\Xinyu\.julia\environments\v1.0\Manifest.toml`
  [7073ff75] ↑ IJulia v1.10.0 ⇒ v1.10.0+ #newzmq (https://github.com/JuliaLang/IJulia.jl.git)
  Building IJulia → `C:\Users\Xinyu\.julia\packages\IJulia\dYgHz\deps\build.log`
```
Works for me. Thanks.

**Comment by stevengj** [comment]
Thanks @XinyuuLi.  Make sure to play around a bit, e.g. try plotting something, to make sure the freeze doesn't occur a little later as it did for other posters in the past.

**Comment by Bob-McCrory** [comment]
I went through the installation steps above, and tried IJulia on the notebook attached below, which was randomly locking up. It contains plots, and it now seems to work reliably. Thank you so much for your efforts.

[LScratch00.zip](https://github.com/JuliaLang/IJulia.jl/files/2347002/LScratch00.zip)

**Comment by jlperla** [comment]
@arnavs Can you try to run th

**Comment by arnavs** [comment]
@jlperla For sure. Will do.

**Comment by jlperla** [comment]
@stevengj I tried to follow the instructions here, but `IJulia#newzmq` branch seems to have been deleted, and the `IJulia#master` doesn't seem to be compatible with the `ZMQ#master`.  The below output is on Julia v1.0.  If there are other instructions for testing, or you prefer people to wait, just advise.

```julia
(scratch) pkg> add ZMQ#master
  Updating git-repo `https://github.com/JuliaInterop/ZMQ.jl.git`
 Resolving package versions...
  Updating `Project.toml`
 [no changes]
  Updating `Manifest.toml`
 [no changes]

(scratch) pkg> add IJulia#master
  Updating git-repo `https://github.com/JuliaLang/IJulia.jl.git`
 Resolving package versions...
ERROR: Unsatisfiable requirements detected for package ZMQ [c2297ded]:
 ZMQ [c2297ded] log:
 ├─possible versions are: 0.6.4 or uninstalled
 ├─ZMQ [c2297ded] is fixed to version 0.6.4+
 └─restricted to versions 1.0.0-* by IJulia [7073ff75] — no versions left
   └─IJulia [7073ff75] log:
     ├─possible versions are: 1.10.0 or uninstalled
     └─IJulia [7073ff75] is fixed to version 1.10.0+
```

**Comment by stevengj** [comment]
@jlperla, now that the REQUIRE file of IJulia lists ZMQ 1.0, I think you may have to manually check out the `master` branch until the new ZMQ is tagged, hopefully later today (JuliaLang/METADATA.jl#17668)

**Comment by stevengj** [comment]
Okay, new releases are tagged.  `Pkg.update` (`up` at the package prompt) should get the new versions in Julia 0.7+.

I'll close this once more people confirm that the issue is fixed (knock on wood).

**Comment by crstnbr** [comment]
It works for me now. Thanks so much for fixing this!

**Comment by zhangzw16** [comment]
I came up with the freeze problem days ago. My Julia version is 1.0.0, and I'm using windows10.
After I tried

```
Pkg.add("ZMQ")
Pkg.add("IJulia")
Pkg.update()
```
the jupyter notebook now works fine and don't freeze! Seem to solved the problem.


Besides, I have tried this before
```
(v0.7) pkg> add IJulia#master
ENV["IJULIA_DEBUG"] = true
(v0.7) pkg> build IJulia
```
and it didn't solve the problem.

**Comment by cwb0** [comment]
I've no success: I did the following:

> julia> versioninfo()
Julia Version 0.7.0
Commit a4cb80f3ed (2018-08-08 06:46 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, haswell)
Environment:
  JULIA_EDITOR = "C:\Users\Carsten\AppData\Local\atom\app-1.30.0\atom.exe" -a
  JULIA_NUM_THREADS = 2

julia> using ZMQ

julia> using IJulia

julia> ENV["IJULIA_DEBUG"] = true
true

(v0.7) pkg>  build IJulia
  Building Conda ──→ `C:\Users\Carsten\.julia\packages\Conda\m7vem\deps\build.log`
  Building MbedTLS → `C:\Users\Carsten\.julia\packages\MbedTLS\Qo8TN\deps\build.log`
  Building ZMQ ────→ `C:\Users\Carsten\.julia\packages\ZMQ\ABGOx\deps\build.log`
  Building IJulia ─→ `C:\Users\Carsten\.julia\dev\IJulia\deps\build.log`
 Resolving package versions...

julia> IJulia.notebook()

Within notebook I chose a new worksheet (Julia 0.7.0). The process 

> Kernel  starting, please....  

starts but eventually I've got the message 

> Dead Kernel

**Comment by fredrikekre** [comment]
What versions of ZMQ and IJulia are you using? (Whats the output of `using Pkg; Pkg.status()`?)

**Comment by cwb0** [comment]
(v0.7) pkg> st
    Status `C:\Users\Carsten\.julia\environments\v0.7\Project.toml`
  [c52e3926] Atom v0.7.5
  [0c46a032] DifferentialEquations v5.2.1
  [7a1cc6ca] FFTW v0.2.4
  [5789e2e9] FileIO v1.0.1
  [7073ff75] IJulia v1.11.1+ [`C:\Users\Carsten\.julia\dev\IJulia`]
  [033835bb] JLD2 v0.1.1
  [e5e0dc1b] Juno v0.5.2
  [b964fa9f] LaTeXStrings v1.0.2
  [5fb14364] OhMyREPL v0.3.0
  [d96e819e] Parameters v0.10.0
  [91a5bcdd] Plots v0.19.3
  [92933f4c] ProgressMeter v0.6.0
  [438e738f] PyCall v1.18.4
  [d330b81b] PyPlot v2.6.3
  [c2297ded] ZMQ v1.0.0
  [ade2ca70] Dates
  [37e2e46d] LinearAlgebra
  [de0858da] Printf
  [10745b16] Statistics

**Comment by cwb0** [comment]
now it seems to be solved. Thank you for your prompt help

**Comment by Bob-McCrory** [comment]
fredrikekre's solution above also solved my problem, which was manifested by the kernel repeatedly dying with a message: ZMQ socket has no field data.

Thank you.
